src/main/scala/abalone/variant/Variant.scala

```diff
@@ -102,17 +102,80 @@ abstract class Variant private[variant] (
         )
     }.toMap
 
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
+    def generateMove(orig: Pos, dest: Pos, category: String): Some[(String, Move)] =
+      Some( (category, Move(Piece(situation.player, Role.defaultRole), orig, dest, situation, boardAfter(situation, orig, dest), true, if (category == "pushout") Some(dest) else None)) )
+
+    def generateSideMoves(lineOfMarbles: List[Pos], direction: Option[String]): List[(String, Move)] = {
+      List(
+        if (lineOfMarbles.size == 3) generateSideMoves(List(lineOfMarbles(0), lineOfMarbles(1)), direction)
+        else None,
+        if (!lineOfMarbles.map(canLeftSideMove(situation, _, direction)).contains(false))
+          generateMove(lineOfMarbles(0), lineOfMarbles.last.sideMovesDirsFromDir(direction)._1.get, "side")
```

Comment:
I'm still having .get

Reply:
If ever you want to use `.get` this means that you might have `None`, and you'll want to handle the case where you have `None`. Writing `.get` allows you to compile, but if this ever resolves to `None` at run time you will get a run time error which you haven't handled.
So instead you need to `map` on the `Option` which allows you to have the concrete value inside the `map` and ensures this code never runs if you have `None`. 
I also don't like `last` or `lineOfMarbles(0)` because these also fail at run time on an empty list (I mentioned `dropRight` above and this is safe because `List().dropRight(1)` == `List()`.
So I would use `lineOfMarbles.headOption` which gives you an `Option` of the 0th element and `None` if `lineOfMarbles.isEmpty`. For `.last` I use `.reverse.headOption`.

This means you want to use a combination of map and flatMap chained together to call `generateMove` safely.
Consider this simplified example:
```
scala> def myFunc(a: Int, b: Int): Option[Int] = Some(a+b) //this is representing generateMove
def myFunc(a: Int, b: Int): Option[Int]

scala> val myList = List(11,12,13)
val myList: List[Int] = List(11, 12, 13)

scala> def intToOpt(i: Int): Option[Int] = Some(i) //this is representing sideMovesDirsFromDir(direction)._1
def intToOpt(i: Int): Option[Int]

scala> myFunc(myList(0), intToOpt(myList.last).get) //this is what you are currently doing
val res11: Option[Int] = Some(24)

scala> myList.headOption.flatMap(h => myFunc(h, intToOpt(myList.last).get)) //handling lineOfMarbles(0) safely
val res15: Option[Int] = Some(24)

scala> myList.headOption.flatMap{h => myList.reverse.headOption.flatMap{t => myFunc(h, intToOpt(t).get)}} //handling lineOfMarbles.last safely
val res16: Option[Int] = Some(24)

scala> myList.headOption.flatMap{h => myList.reverse.headOption.flatMap{t => intToOpt(t).flatMap{t2 => myFunc(h, t2)}}} //everything safe now, intToOpt handled.
val res17: Option[Int] = Some(24)
```
Now if `myList` was empty this is the difference:
```
scala> val myEmptyList: List[Int] = List()
val myEmptyList: List[Int] = List()

scala> myEmptyList.headOption.flatMap{h => myEmptyList.reverse.headOption.flatMap{t => intToOpt(t).flatMap{t2 => myFunc(h, t2)}}}
val res18: Option[Int] = None

scala> myFunc(myEmptyList(0), intToOpt(myEmptyList.last).get)
java.lang.IndexOutOfBoundsException: 0
  at scala.collection.LinearSeqOps.apply(LinearSeq.scala:117)
  at scala.collection.LinearSeqOps.apply$(LinearSeq.scala:114)
  at scala.collection.immutable.List.apply(List.scala:79)
  ... 59 elided
```